mm2src/mm2_main/src/heartbeat_event.rs

```diff
@@ -1,52 +1,55 @@
 use async_trait::async_trait;
-use common::{executor::{SpawnFuture, Timer},
-             log::info};
-use futures::channel::oneshot::{self, Receiver, Sender};
-use mm2_core::mm_ctx::MmArc;
-use mm2_event_stream::{behaviour::{EventBehaviour, EventInitStatus},
-                       Event, EventName, EventStreamConfiguration};
+use common::executor::Timer;
+use futures::channel::oneshot;
+use mm2_event_stream::{Broadcaster, Event, EventStreamer, NoDataIn, StreamHandlerInput};
+use serde::Deserialize;
+use serde_json::Value as Json;
+
+#[derive(Deserialize)]
+#[serde(deny_unknown_fields, default)]
+struct HeartbeatEventConfig {
+    /// The time in seconds to wait before sending another ping event.
+    pub stream_interval_seconds: f64,
+}
+
+impl Default for HeartbeatEventConfig {
+    fn default() -> Self {
+        Self {
+            stream_interval_seconds: 5.0,
+        }
+    }
+}
 
 pub struct HeartbeatEvent {
-    ctx: MmArc,
+    config: HeartbeatEventConfig,
 }
 
 impl HeartbeatEvent {
-    pub fn new(ctx: MmArc) -> Self { Self { ctx } }
+    pub fn try_new(config: Option<Json>) -> serde_json::Result<Self> {
+        Ok(Self {
+            config: config.map(serde_json::from_value).unwrap_or(Ok(Default::default()))?,
+        })
+    }
 }
 
 #[async_trait]
-impl EventBehaviour for HeartbeatEvent {
-    fn event_name() -> EventName { EventName::HEARTBEAT }
+impl EventStreamer for HeartbeatEvent {
+    type DataInType = NoDataIn;
 
-    async fn handle(self, interval: f64, tx: oneshot::Sender<EventInitStatus>) {
-        tx.send(EventInitStatus::Success).unwrap();
+    fn streamer_id(&self) -> String { "HEARTBEAT".to_string() }
 
-        loop {
-            self.ctx
-                .stream_channel_controller
-                .broadcast(Event::new(Self::event_name().to_string(), json!({}).to_string()))
-                .await;
+    async fn handle(
+        self,
+        broadcaster: Broadcaster,
+        ready_tx: oneshot::Sender<Result<(), String>>,
+        _: impl StreamHandlerInput<NoDataIn>,
+    ) {
+        ready_tx.send(Ok(())).unwrap();
 
-            Timer::sleep(interval).await;
-        }
-    }
+        loop {
+            broadcaster.broadcast(Event::new(self.streamer_id(), json!({})));
```

Comment:
I think we should use `broadcast_all` and remove `heartbeat::enable` RPC, so when any client subscribes to any other streamer, they should start receiving heartbeat events. Either this or we advise anyone using the streaming interface to subscribe to heartbeat events on start as this is needed for health check that streaming is working fine. I think the first approach is better and if we can send a final event (different one) before a streamer removes a client so that the client knows it's not listening to anything and there is no heartbeat events is a good approach. You can do this in next PR as merging without it is fine.

Reply:
to be covered later on https://github.com/KomodoPlatform/komodo-defi-framework/issues/2207